src/sdk/namespace_fs.js

```diff
@@ -536,15 +679,31 @@ class NamespaceFS {
                 common_prefixes: [],
                 is_truncated,
                 next_marker: undefined,
+                next_version_id_marker: undefined,
             };
             for (const r of results) {
+                let obj_info;
                 if (r.common_prefix) {
                     res.common_prefixes.push(r.key);
                 } else {
-                    res.objects.push(this._get_object_info(bucket, r.key, r.stat, 'null'));
+                    obj_info = this._get_object_info(bucket, r.key, r.stat, 'null', true);
+                    if (!list_versions && obj_info.delete_marker) {
+                        continue;
+                    }
+                    if (this._is_hidden_version_path(obj_info.key)) {
+                        obj_info.key = path.normalize(obj_info.key.replace(HIDDEN_VERSIONS_PATH + '/', ''));
+                        obj_info.key = obj_info.key.substring(0, obj_info.key.indexOf('_mtime-'));
+                    }
+                    res.objects.push(obj_info);
                 }
                 if (res.is_truncated) {
-                    res.next_marker = r.key;
+                    if (list_versions && r.key.includes('_mtime-')) {
+                        const next_version_id_marker = r.key.substring(r.key.lastIndexOf('/') + 1);
+                        res.next_version_id_marker = next_version_id_marker;
+                        res.next_marker = key_marker;
```

Comment:
why the next_marker is the key_marker we got as parameter?

Reply:
I think I need to fetch the latest_key name and assign to next_key here

`res.next_marker = next_version_id_marker.substring(0, next_version_id_marker.indexOf('_mtime'));`